SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestNetReference/SpatialTestNetReference.cpp

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestNetReference.h"
+#include "GameFramework/PlayerController.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMovement/TestMovementCharacter.h"
+#include "Kismet/GameplayStatics.h"
+#include "CubeWithReferences.h"
+#include "SpatialGDKSettings.h"
+
+
+/**
+ * This test automates the Net Reference Test gym, which tested that references to replicated actors are stable when actors go in and out of relevance.
+ * This test also adds an interest check on top of the previously mentioned Gym.
+ * 
+ * The test includes a single server and two client workers. For performance considerations, the only client that is executing the test is Client 1.
+ * The flow is as follows:
+ * - Setup:
+ *   - The Server spawns 4 CubeWithReferences objects and sets up their references.
+ * - Test:
+ *	 - The test contains 2 runs of the same flow: Client 1 moves its character to 4 specific locations and, after arriving at each location, it checks that:
+ *			1) The correct amount of cubes are present in the world, based on the default NetCullDistanceSquared of the PlayerController.
+ *			2) The references to the replicated actors are correct.
+ * - Clean-up:
+ *	- The Server destroys the previously spawned CubeWithReferences
+ *  - The test variables are reset before the test finishes so that it can be ran multiple times.
+ */
+
+ASpatialTestNetReference::ASpatialTestNetReference()
+	: Super()
+{
+	Author = "Andrei";
+	Description = TEXT("Test Net Reference");
+
+	LocationIndex = 0;
+	TimerHelper = 0.0f;
+
+	TestLocations.Add(TPair<FVector, int> (FVector(0.0f, -15000.0f, 40.0f), 1));
```

Comment:
These seem quite large, can you still see them when executing the test - for ease of debugging?

Reply:
No you cannot, but this is because I am basing the test on the default NetCullDistanceSquared and this forces me to have the cubes far away and move the character to this kind of locations. I will try thinking of a way of making debugging easier.

Reply:
Perhaps override NetCull distance for a base TestActor such that it can be easily visible? However since this is a gym I believe it's not that important.
One improvement I would suggest would be to get this these distances used to move out of the `NetCullDistanceSquared` range set in some constants class to make them re-usable and easily understandable. The bare minimum would be some comments on the distances used in regards to the `NetCullDistanceSquared`.

Reply:
In regards to overriding the NetCullDistance for the TestActor, I believe you will actually need to do this for the default Pawn perhaps?

Reply:
This is an automated test, if you want to see a similar situation but with overriden distances, there still is the Gym.

Reply:
I have added some support for visual debugging by scaling the cubes and also moving the camera so that it can always see all the 4 cubes. If you comment the line that updates the PositionUpdateFrequency, I think debugging is relatively good.